File: control/fleet_gateway/test/test_map_annotations.js
const test = require("node:test");
const assert = require("node:assert/strict");

const annotations = require("../web/map_annotations.js");

test("virtual wall includes its configured robot safety margin", () => {
  const wall = {
    enabled: true,
    type: "virtual_wall",
    name: "복도 벽",
    width_m: 0.08,
    safety_margin_m: 0.16,
    points: [{ x: 0, y: -1 }, { x: 0, y: 1 }],
  };

  assert.equal(annotations.blocksPoint(wall, 0.19, 0), true);
  assert.equal(annotations.blocksPoint(wall, 0.21, 0), false);
  assert.match(annotations.blockedReason([wall], 0, 0), /복도 벽/);
});

test("privacy and keepout polygons block their interior and margin", () => {
  const privacy = {
    enabled: true,
    type: "privacy",
    name: "상담실",
    safety_margin_m: 0.1,
    points: [
      { x: 0, y: 0 },
      { x: 1, y: 0 },
      { x: 1, y: 1 },
      { x: 0, y: 1 },
    ],
  };

  assert.equal(annotations.blocksPoint(privacy, 0.5, 0.5), true);
  assert.equal(annotations.blocksPoint(privacy, 1.05, 0.5), true);
  assert.equal(annotations.blocksPoint(privacy, 1.2, 0.5), false);
});

test("charging positions are visible destinations, not barriers", () => {
  const charging = {
    enabled: true,
    type: "charging",
    name: "충전기",
    pose: { x: 0.5, y: 0.5, yaw: 0 },
  };

  assert.equal(annotations.blocksPoint(charging, 0.5, 0.5), false);
  assert.equal(annotations.blockedReason([charging], 0.5, 0.5), "");
});

File: control/fleet_gateway/test/test_map_annotations.py
import json

import pytest

from fleet_gateway.map_annotations import (
    MapAnnotationStore,
    annotation_blocks_point,
    compile_keepout_mask,
)


def occupancy_map():
    return {
        "frame_id": "map",
        "width": 20,
        "height": 20,
        "resolution": 0.1,
        "origin": {"x": 0.0, "y": 0.0, "yaw": 0.0},
        "data": [0] * 400,
    }


def test_store_persists_hard_policies_and_charging_destinations(tmp_path):
    path = tmp_path / "annotations.json"
    store = MapAnnotationStore(path)
    wall = store.create(
        "tb1",
        {
            "type": "virtual_wall",
            "name": "복도 벽",
            "points": [{"x": 1.0, "y": 0.2}, {"x": 1.0, "y": 1.8}],
            "width_m": 0.08,
            "safety_margin_m": 0.16,
        },
        occupancy_map(),
    )
    charging = store.create(
        "tb1",
        {
            "type": "charging",
            "name": "충전기",
            "pose": {"x": 0.3, "y": 0.3, "yaw": 1.57},
        },
        occupancy_map(),
    )

    reloaded = MapAnnotationStore(path).list("tb1")

    assert [item["annotation_id"] for item in reloaded] == [
        wall["annotation_id"],
        charging["annotation_id"],
    ]
    assert annotation_blocks_point(wall, 1.19, 1.0)
    assert not annotation_blocks_point(charging, 0.3, 0.3)
    assert json.loads(path.read_text(encoding="utf-8"))["version"] == 1


def test_polygon_blocks_interior_margin_and_mask_cells():
    store = MapAnnotationStore()
    privacy = store.create(
        "tb1",
        {
            "type": "privacy",
            "name": "상담 구역",
            "points": [
                {"x": 0.5, "y": 0.5},
                {"x": 1.5, "y": 0.5},
                {"x": 1.5, "y": 1.5},
                {"x": 0.5, "y": 1.5},
            ],
            "safety_margin_m": 0.1,
        },
        occupancy_map(),
    )

    mask = compile_keepout_mask(occupancy_map(), [privacy])

    assert store.blocked_reason("tb1", 1.0, 1.0)
    assert store.blocked_reason("tb1", 1.55, 1.0)
    assert not store.blocked_reason("tb1", 1.8, 1.0)
    assert mask["data"][10 * 20 + 10] == 100
    assert mask["data"][1 * 20 + 1] == 0


def test_rejects_geometry_outside_map_and_policy_that_traps_robot():
    store = MapAnnotationStore()
    with pytest.raises(ValueError, match="inside the map"):
        store.create(
            "tb1",
            {
                "type": "virtual_wall",
                "points": [{"x": -0.1, "y": 0.2}, {"x": 1.0, "y": 0.2}],
            },
            occupancy_map(),
        )

    with pytest.raises(ValueError, match="trap the robot"):
        store.create(
            "tb1",
            {
                "type": "keepout",
                "points": [
                    {"x": 0.0, "y": 0.0},
                    {"x": 0.8, "y": 0.0},
                    {"x": 0.8, "y": 0.8},
                    {"x": 0.0, "y": 0.8},
                ],
            },
            occupancy_map(),
            protected_pose={"x": 0.4, "y": 0.4, "yaw": 0.0},
        )
